В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [12]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [13]:
alpha = 0.05
null_mean = 170

In [14]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [20]:
print("Середнє вибірки: {:.2f}".format(np.mean(heights)))
print("Стандартне відхилення вибірки: {:.2f}".format(np.std(heights)))
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='larger')
print("ztest_Score: ", ztest_Score, "p_value: ", p_value)
if p_value < alpha:
    print("Відхилити H0: середній ріст вибірки значно більший за 170 см.")
else:
    print("Прийняти H0: недостатньо доказів, що середній ріст більший за 170 см.")

Середнє вибірки: 170.70
Стандартне відхилення вибірки: 4.65
ztest_Score:  0.8482235371587619 p_value:  0.19815674571091746
Прийняти H0: недостатньо доказів, що середній ріст більший за 170 см.


##### Висновок: 
Хоча середнє вибірки трохи більше за 170, статистичний тест враховує розсіювання даних і розмір вибірки. У цьому випадку розсіювання занадто велике, і вибіркове середнє 170.7 недостатньо відрізняється від 170, щоб вважати це значущим. Тому H0 приймається.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [21]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

Перевіримо чи мають групи хоч якусь різницю у результатах:

In [22]:
ztest(group_a_scores, group_b_scores, value=0)

(np.float64(2.357354864751502), np.float64(0.018405648552528006))

р-значення менше, ніж 0,05, а отже, гіпотеза Н0 відхиляється і ми припускаємо, що групи мають результати. Далі проводимо односторонній тест для визначення чи є новий ефективніший.

In [25]:
ztest_score, p_value = ztest(group_a_scores, group_b_scores, alternative='larger')
print(f"Z-значення: {ztest_score}, P-значення: {p_value}")

if p_value < 0.05:
    print("Відхилити H0: новий курс ефективніший на ", ztest_score, " стандартних відхилення.")
else:
    print("Прийняти H0: недостатньо доказів, що новий курс ефективніший")


Z-значення: 2.357354864751502, P-значення: 0.009202824276264003
Відхилити H0: новий курс ефективніший на  2.357354864751502  стандартних відхилення.


**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [1]:
group_a_scores_1 = [78.55, 72.25, 79.88, 75.  , 76.54]

group_b_scores_1 = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [2]:
import numpy as np
from scipy import stats

In [5]:
stats.ttest_ind(group_a_scores_1, group_b_scores_1, equal_var=False)

TtestResult(statistic=np.float64(1.1226527301193456), pvalue=np.float64(0.3010736990344312), df=np.float64(6.544857383782639))

Тепер p-value>0 і ми приймаємо H0: недостатньо доказів, що новий курс ефективніший.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [10]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [11]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [14]:
ztest_score, p_value = ztest(after_training, before_training, alternative='larger')
print(f"Z-значення: {ztest_score}, P-значення: {p_value}")

if p_value < 0.05:
    print("Відхилити H0: результати кращі на", ztest_score, " стандартних відхилення.")
else:
    print("Прийняти H0: не можна стверджувати, що результати покращились")

Z-значення: 1.3668871347639466, P-значення: 0.08583034256235411
Прийняти H0: не можна стверджувати, що результати покращились


В результаті тесту p-value більше 0.05, що означає, що немає достатніх доказів для відхилення нульової гіпотези. Тобто немає статистично значущої різниці в продуктивності до і після тренінгу на рівні значущості 0.05.